From e190b5233aa3876ef88a5039f69e3fa568694eb6 Mon Sep 17 00:00:00 2001
From: Florian Fuchs <fuchsfl@gmail.com>
Date: Fri, 17 Jul 2026 10:53:57 +0200
Subject: [PATCH] sh: don't copy insns that define a PIC call site label
 [PR89012]

gcc/ChangeLog:

	PR target/89012
	* config/sh/sh.cc (sh_cannot_copy_insn_p): Scan every element of
	a PARALLEL for the call site label, not just the last one.

gcc/testsuite/ChangeLog:

	PR target/89012
	* gcc.target/sh/sh.exp (check_effective_target_sh_fdpic): New.
	* gcc.target/sh/pr89012.c: New test.
---
 gcc/config/sh/sh.cc                   |  9 ++++++---
 gcc/testsuite/gcc.target/sh/pr89012.c | 27 +++++++++++++++++++++++++++
 gcc/testsuite/gcc.target/sh/sh.exp    |  9 +++++++++
 3 files changed, 42 insertions(+), 3 deletions(-)
 create mode 100644 gcc/testsuite/gcc.target/sh/pr89012.c

diff --git a/gcc/config/sh/sh.cc b/gcc/config/sh/sh.cc
index 05b0060..d7a7a67 100644
--- a/gcc/config/sh/sh.cc
+++ b/gcc/config/sh/sh.cc
@@ -2895,9 +2895,12 @@ sh_cannot_copy_insn_p (rtx_insn *insn)
 
   if (TARGET_FDPIC && GET_CODE (pat) == PARALLEL)
     {
-      rtx t = XVECEXP (pat, 0, XVECLEN (pat, 0) - 1);
-      if (GET_CODE (t) == USE && unspec_caller_rtx_p (XEXP (t, 0)))
-	return true;
+      for (int i = 0; i < XVECLEN (pat, 0); i++)
+	{
+	  rtx t = XVECEXP (pat, 0, i);
+	  if (GET_CODE (t) == USE && unspec_caller_rtx_p (XEXP (t, 0)))
+	    return true;
+	}
     }
 
   if (GET_CODE (pat) != SET)
diff --git a/gcc/testsuite/gcc.target/sh/pr89012.c b/gcc/testsuite/gcc.target/sh/pr89012.c
new file mode 100644
index 0000000..8e0aa83
--- /dev/null
+++ b/gcc/testsuite/gcc.target/sh/pr89012.c
@@ -0,0 +1,27 @@
+/* PR target/89012: an sfunc call pattern such as ashlsi3_d_call emits its own
+   PIC call site label via %O, so the insn must not be duplicated.  Here bbro
+   copies the block holding the variable shift, which used to emit .LPCS0:
+   twice and made the assembler fail.  -m2 selects a CPU without dynamic shift
+   instructions, so that the shift becomes an sfunc call in the first place.  */
+/* { dg-do assemble } */
+/* { dg-require-effective-target sh_fdpic } */
+/* { dg-options "-O2 -fpic -m2 -mfdpic" } */
+
+typedef unsigned int vword;
+
+extern void f0 (void);
+extern void f1 (void);
+
+void
+chk (unsigned int n, vword const *v, int c)
+{
+  for (unsigned int i = 0; i < n; i++)
+    {
+      if (c)
+	f0 ();
+      else
+	f1 ();
+      if (v[i / 32] & (1u << (i % 32)))
+	f0 ();
+    }
+}
diff --git a/gcc/testsuite/gcc.target/sh/sh.exp b/gcc/testsuite/gcc.target/sh/sh.exp
index 97172f5..8415a2b 100644
--- a/gcc/testsuite/gcc.target/sh/sh.exp
+++ b/gcc/testsuite/gcc.target/sh/sh.exp
@@ -24,6 +24,15 @@ if ![istarget sh*-*-*] then {
 # Load support procs.
 load_lib gcc-dg.exp
 
+# Return 1 if the target supports -mfdpic
+proc check_effective_target_sh_fdpic { } {
+    return [check_no_compiler_messages sh_fdpic object {
+	     #ifndef __SH_FDPIC__
+	     #error ""
+	     #endif
+    } "-mfdpic"]
+}
+
 # Return 1 if target is SH2A
 proc check_effective_target_sh2a { } {
     return [check_no_compiler_messages sh2a object {
